velox/common/memory/HashStringAllocator.h

```diff
@@ -555,29 +555,45 @@ struct AlignedStlAllocator {
   }
 
   T* FOLLY_NONNULL allocate(std::size_t n) {
-    if (n * sizeof(T) > HashStringAllocator::kMaxAlloc && poolAligned_) {
-      return reinterpret_cast<T*>(allocator_->allocateFromPool(n * sizeof(T)));
+    if (n * sizeof(T) > HashStringAllocator::kMaxAlloc) {
+      if (poolAligned_) {
+        return reinterpret_cast<T*>(
+            allocator_->allocateFromPool(n * sizeof(T)));
+      } else {
+        // Memory layout:
+        // - Header
+        // - padding(length is stored in `delta`)
+        // - delta(4 bytes storing the size of padding)
+        // - the aligned ptr
+        auto paddedSize =
+            checkedPlus<size_t>(Alignment + 4, checkedMultiply(n, sizeof(T)));
+        // Allocate the memory block with a Header.
+        auto header = reinterpret_cast<HashStringAllocator::Header*>(
+            allocator_->allocateFromPool(
+                paddedSize + sizeof(HashStringAllocator::Header)));
+        // Init the header with correct size info.
+        new (header) HashStringAllocator::Header(paddedSize);
+        auto ptr = header->begin();
+
+        return alignPtr((char*)ptr, n, paddedSize);
+      }
     }
+
     // Allocate extra Alignment bytes for alignment and 4 bytes to store the
     // delta between unaligned and aligned pointers.
-    auto size =
+    // Memory layout:
+    // - padding(length is stored in `delta`)
+    // - delta(4 bytes storing the size of padding)
+    // - the aligned ptr
+    auto paddedSize =
         checkedPlus<size_t>(Alignment + 4, checkedMultiply(n, sizeof(T)));
-    auto ptr = reinterpret_cast<T*>(allocator_->allocate(size)->begin());
-
-    // Align 'ptr + 4'.
-    void* alignedPtr = (char*)ptr + 4;
-    size -= 4;
-    std::align(Alignment, n * sizeof(T), alignedPtr, size);
+    auto ptr = reinterpret_cast<T*>(allocator_->allocate(paddedSize)->begin());
 
-    // Write alignment delta just before the aligned pointer.
-    int32_t delta = (char*)alignedPtr - (char*)ptr - 4;
-    *reinterpret_cast<int32_t*>((char*)alignedPtr - 4) = delta;
-
-    return reinterpret_cast<T*>(alignedPtr);
+    return alignPtr((char*)ptr, n, paddedSize);
   }
 
   void deallocate(T* FOLLY_NONNULL p, std::size_t n) noexcept {
-    if (n * sizeof(T) > HashStringAllocator::kMaxAlloc) {
+    if (n * sizeof(T) > HashStringAllocator::kMaxAlloc && poolAligned_) {
```

Comment:
Let's use `freeToPool` also in case of `!poolAligned_` to be symmetric

Reply:
`HashStringAllocator::free` can automatically handle a large allocation prefixed by a `Header`:

https://github.com/facebookincubator/velox/blob/3020196b001130a9027b09d6b3b95385b90bc2cb/velox/common/memory/HashStringAllocator.cpp#L433-L436